### PR TITLE
Assign 'ferie' color style the highest priority

### LINKED ISSUE #567: Calendrier : une cellule pour un jour à la fois "férié" et "congé" prend la couleur d'un jour de congé
Est-ce le comportement souhaité ?
Ça semble dû juste à la priorité dans les règles CSS, la déclaration ".conge_all.conge_ok" étant plus précise et donc prioritaire sur ".ferie".

Ce qui se complique pour mon cas précis, c'est que nous avons une règle interne particulière au boulot qui fait qu'un jour férié peut parfois être décompté comme un jour de congé et donc qu'on devrait avoir visuellement un jour de congé pour les jours fériés dans cette situation...

### Système et application

* Système d'exploitation du serveur : Ubuntu 16.04
* Serveur HTTP et version : Apache 2.4.18
* Version MySQL et PHP : MySql 5.7.20, PHP 7.0
* Navigateur et version : Chrome 63
* Version de l'application : 1.10.1 (kantra)
* Avez-vous des notifications dans la console du navigateur : non
* Avez-vous des erreurs dans le log du serveur http (ex : /var/log/apache2) : non
* Méthodes de connexion [interne / ldap / sso] : ldap

### Reproduction du bug

* Prérequis : avoir un jour férié pendant une période de congés

### PR DESCRIPTION
**Pull request by benrubson**
Hi,

This PR closes #567.

A 'ferie' day now has its correct color assigned :

screen shot 2018-10-04 at 09 28 01

Thank you 👍

### CODE REVIEW

**Comment by benrubson** [comment]
I'm not sure why _Better Code Hub_ asks for modifications, I don't have access to its worklog.

**Comment by wouldsmina** [comment]
thanks @benrubson. BCH says :  This code needs to be refactored 
it's for the entire calendrier.php file... don't worry.

**Comment by benrubson** [comment]
Nice, thank you @wouldsmina 👍

File: calendrier.php [hunk 1]
```diff
@@ -28,7 +28,7 @@ function getUrlMois(\DateTimeInterface $date, $idGroupe)
 
 function getClassesJour(\App\Libraries\Calendrier\Evenements $evenements, $nom, $jour, \DateTimeInterface $moisDemande)
 {
-    return implode(' ', $evenements->getEvenementsDate($nom, $jour));
+    return preg_replace('/(^|.* )ferie( .*|$)/', 'ferie', implode(' ', $evenements->getEvenementsDate($nom, $jour)));
```
[the remark this answers is not in the document]
Comment (benrubson) [reply calendrier.php]: Thank you for your feedback @prytoegrian 👍 

> I would prefer asserting a priority of 'ferie' event and not set an event if there's already 'ferie' for this day and this employee. I have more or less already did it in https://github.com/libertempo/web/blob/develop/App/Libraries/Calendrier/Evenements.php#L136L145

Something like the second commit ?

**Comment by benrubson** [comment]
Thank you for your feedback @prytoegrian 👍 

> I would prefer asserting a priority of 'ferie' event and not set an event if there's already 'ferie' for this day and this employee. I have more or less already did it in https://github.com/libertempo/web/blob/develop/App/Libraries/Calendrier/Evenements.php#L136L145

Something like the second commit ?

File: calendrier.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (prytoegrian) [reply calendrier.php]: Exactly ! I'll test that.

**Comment by prytoegrian** [comment]
:warning: Units tests don't pass anymore, look the travis check. If you have any problem to understand how to correct, let's discuss it.

**Comment by benrubson** [comment]
You're right, unit tests are failing, _2017-02-12_ is seen as a _ferie_ day instead of a _weekend_ day.
I'm not sure however where _ferie_ is wrongly set in my commit.
Did you already catch the bug ?

**Comment by prytoegrian** [comment]
None of yours commits is wrong... not directly, say. I asserted some fact in the unit tests about the behavior tested and you've changed this behavior.

In testGetEvenenementsDateWeekend test function, I check that « weekend » is really a absorbent element, forbidding that a date will be « weekend » and « conges » at the same time. To do that, I mock the calling of `$this->evenement->getListe()` : 
```php
// Tests/Units/App/Libraries/Calendrier/Evenements.php::testGetEvenenementsDateWeekend
{
        // définition d'un autre événement avec weekend
        $this->calling($this->evenement)->getListe = ['2017-02-12'];
        $this->calling($this->evenement)->getListe[4] = [];
        // ....
}
```
[Atoum](http://docs.atoum.org/en/latest/mocking_systems.html#modify-the-behaviour-of-a-mock) proposes a syntax for the mocking of functions : 
* $this->calling($this->mock)->function[1] = ''; // for the very first call of the function
* $this->calling($this->mock)->function[2] = ''; // for the second one
* and so on

There's a fallback though. If you didn't set an offset, it means « *all (others) calls* ». Here, I mixed up theses syntaxes so that both `fetchWeekends()` **and** `fetchFeries()` return `['2017-02-12']`. Since these two types both are absorbent elements now, « ferie » erases « weekend ».

To me, the correction is not very difficult, you already know all of the needed. 

If you're on Unix, look at the Makefile and particularly `make test`.

**Comment by benrubson** [comment]
Thank you for your detailed feedback @prytoegrian 👍
I pushed a fix (so that `2017-02-12` will not be seen as `ferie`) but Travis seems to be stuck for now.

**Comment by prytoegrian** [comment]
I changed travis settings, I want to try something. Let me check tests myself, and I'll merge today.

**Comment by benrubson** [comment]
OK :)
I think this commit will do the trick, however I did not tested it (I was relying on Travis for this), so I'm sorry if it still fails :-/

**Comment by prytoegrian** [comment]
Tested (in both ways). Thanks a lot !

**Comment by benrubson** [comment]
Many thanks too for your help 👍

**Pull request merged**